pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java

```diff
@@ -214,31 +215,19 @@ private String getMultiStageQueryResponse(String query, String queryOptions, Htt
       // tenants can be completely disjoint. The leaf stages which access segments will be processed on the respective
       // server tenants for each table. The intermediate stages can be processed in either or all of the server tenants
       // belonging to the tables.
-      brokerTenant = getCommonBrokerTenant(tableConfigList);
-      if (brokerTenant == null) {
+      brokerTenants = getCommonBrokerTenants(tableConfigList);
+      if (brokerTenants.isEmpty()) {
         return QueryException.getException(QueryException.BROKER_REQUEST_SEND_ERROR, new Exception(
             String.format("Unable to dispatch multistage query with multiple tables : %s " + "on different tenant",
                 tableNames))).toString();
       }
     } else {
       // TODO fail these queries going forward. Added this logic to take care of tautologies like BETWEEN 0 and -1.
-      List<String> allBrokerList = new ArrayList<>(_pinotHelixResourceManager.getAllBrokerTenantNames());
-      brokerTenant = allBrokerList.get(RANDOM.nextInt(allBrokerList.size()));
+      brokerTenants = _pinotHelixResourceManager.getAllBrokerTenantNames();
       LOGGER.error("Unable to find table name from SQL {} thus dispatching to random broker.", query);
     }
-    List<String> instanceIds = new ArrayList<>(_pinotHelixResourceManager.getAllInstancesForBrokerTenant(brokerTenant));
-
-    if (instanceIds.isEmpty()) {
-      return QueryException.BROKER_RESOURCE_MISSING_ERROR.toString();
-    }
-
-    instanceIds.retainAll(_pinotHelixResourceManager.getOnlineInstanceList());
-    if (instanceIds.isEmpty()) {
-      return QueryException.BROKER_INSTANCE_MISSING_ERROR.toString();
-    }
-
-    // Send query to a random broker.
-    String instanceId = instanceIds.get(RANDOM.nextInt(instanceIds.size()));
+    List<String> instanceIds = findCommonBrokerInstance(brokerTenants);
+    String instanceId = selectInstanceId(instanceIds);
```

Comment:
So previous behaviour was that we selected one broker at random and sent query there. Now we are trying to find a common instance among all the brokers. This will result in not sending the query at all right? Correct me if I am getting this wrong. 

What i understand from this change is say there are 3 brokers: broker1, broker2, broker3. broker1 and broker2 has one instance in common and broker3 has no intersection. Then we should send the query to the intersection of broker1 and broker2. But here we will fail the query as all 3 of them have no intersection.

Reply:
no. i think that's the problem with previous implementation. what i did here is:
1. get a UNION of all the table's broker tag from all the tables (not intersect). e.g. tableA has tenant tag A and tableB has tenant tag B then the UNION will be [A, B]
2. get a list of brokers that contains all the broker tags in the UNION --> only the broker that are tagged with both A and B will be selected. 
3. select a random broker from the list of brokers.

Reply:
Umm yes but say in a scenario where there are 3 tables  -- tableA, tableB, tableC on broker tenants A, B, C.
Interesection of [A, B] has 2 broker-instances common but brokertenant C as no instances common to either A or B. In this case, `findCommonBrokerInstance` will return Collections.emptyList() and `selectInstanceId` will return BROKER_RESOURCE_MISSING_ERROR.

Here `_pinotHelixResourceManager.getAllBrokerTenantNames()` doesn't include tableNames that are passed in the query but all tables in the cluster. Chances are high that there are no common instances among all these table brokerTenants.

Reply:
well if your query access table A B and C at the same time it should do just that. b/c there's no way to run a query to join A B and C together without a broker knowing the info of all 3 tables

Reply:
But what if my query access only table A and B here even then it would fail right?

Reply:
no it won't. the brokerTenant (brokerTenantUnion in new code) will only contain the unioned tag from table A and B.